File: frontend/src/app/models/devices.ts
export interface DeviceFind {
  expression: Record<string, unknown> | string;
  flags: string; // default: ""
}

export interface DeviceFindByIP {
  ipAddress: string;
  ipStatus: string;  // default: ""
  flags: string;     // default: ""
}

export interface DeviceSetName {
  name: string;
  source: string;
}

export interface DeviceSetType {
  type: string;
  source: string;
}

export interface DeviceTagRequest {
  expression: string; // default: ""
  traverse: string;   // default: ""
}

export interface DeviceAlternativeRequest {
  alternative: string;
}

export interface DeviceAlternativeRules {
  rules: unknown[];
}

File: backend/models/devices.py
from typing import Any

from pydantic import BaseModel


class DeviceFind(BaseModel):
    expression: dict | str
    flags: str = ""


class DeviceFindByIP(BaseModel):
    ipaddress: str
    ipstatus: str = ""
    flags: str = ""


class DeviceSetName(BaseModel):
    name: str
    source: str


class DeviceSetType(BaseModel):
    type: str
    source: str


class DeviceTagRequest(BaseModel):
    expression: str = ""
    traverse: str = ""


class DeviceAlternativeRequest(BaseModel):
    alternative: str


class DeviceAlternativeRules(BaseModel):
    rules: list[Any]
